### DISCUSSION TITLE
Update Dashboard

### DISCUSSION THREAD

**Discussion by adhossain** [davetcc/tcMenu#442] [Q&A] [answered]
Hi again, I am using ESP8266, SDA1306 OLED diplay. I can takeover the display and giveback to tcMenu when needed. I want to use a sonor sensor to display the value on the dashboard. Usually we would put that code in the Loop so that the display updates continuously. But I am not sure how to implement the same using tcMenu framework.

Looking forward for your kind support.

**Comment by davetcc** [comment, marked as answer]
You have a few possibilities here, from taking over the display and doing your own drawing, to using a dashboard.

Take a look at the examples, they show all the various ways of handling display take-over. In short, instead of writing your code in loop, you'd write your code in the callback you setup with takeOverDisplay when the renderer calls you. Many of the examples show off this functionality, here are a few that spring to mind:

In [stm32DuinoDemo](https://github.com/davetcc/tcMenuLib/tree/master/examples/arduino32/stm32DuinoDemo) it uses the object-oriented way of working.

In [analogJoystick1306Ascii](https://github.com/davetcc/tcMenuLib/tree/master/examples/avr/analogJoystick1306Ascii) it uses a callback function that is called frequently to to the drawing.

Lastly, in [picoAdafruitDashboard](https://github.com/davetcc/tcMenuLib/tree/master/examples/arduino32/picoAdafruitDashboard) it uses the dashboard approach.

**Comment by adhossain** [comment]
Hi @davetcc ,
I ran into a new problem. I want to use A02YYUW Waterproof Ultrasonic Sensor with ESP32. I connected the sensor using the hardware serial pin of ESP 32 and display the value on SDA1306 OLED display. The "ReadUltrasonic" section reads ultrasonic sensor value and print to serial. The problem is when I call "ReadUltrasonic" inside "ShowHome", the reading stops. I only receive the value for once.

Please note; ShowHome is actually the homescreen/ dashboard which I want to display on boot or when the menu is idle/ exit from menu.

The code given below;

`#include "Water_Tank_Multi_Mode_menu.h"
#include <Arduino.h>
#include <HardwareSerial.h>
HardwareSerial Ultrasonic_Sensor(2);  //   TX2 (pin 17), RX2 (pin 16)

unsigned char data[4] = {};
float distance;


bool homeShowing;

void setup() {
  EEPROM.begin(512);
  Ultrasonic_Sensor.begin(9600);  // Initialize the hardware serial
  Serial.begin(115200);
  delay(2000);
  setupMenu();
  menuMgr.load();
  renderer.takeOverDisplay(ShowHome);
}

void loop() {
  taskManager.runLoop();
}


void CALLBACK_FUNCTION saveSettingsExit(int id) {
  // TODO - your menu change code
  menuMgr.save();
  EEPROM.commit();
  renderer.takeOverDisplay(ShowHome);
}

void ReadUltrasonic() {
  do {
    for (int i = 0; i < 4; i++) {
      data[i] = Ultrasonic_Sensor.read();
    }
  } while (Ultrasonic_Sensor.read() == 0xff);

  Ultrasonic_Sensor.flush();

  if (data[0] == 0xff) {
    int sum;
    sum = (data[0] + data[1] + data[2]) & 0x00FF;
    if (sum == data[3]) {
      distance = (data[1] << 8) + data[2];
      Serial.print("distance=");
      distance = distance / 10;
      Serial.println(distance);
    }
  }
  delay(100);
}

void ShowHome(unsigned int value, RenderPressMode mode) {

  gfx.clearBuffer();                   // clear the internal memory
  gfx.setFont(u8g2_font_ncenB08_tr);   // choose a suitable font
  gfx.drawStr(20, 10, "Water Level");  // write something to the internal memory
  //gfx.setFont(u8g2_font_luRS19_tr);  // choose a suitable font
  ReadUltrasonic();
  gfx.sendBuffer();  // transfer internal memory to the display
  delay(1000);
}`

**Reply by davetcc** [reply]
This discussion board only really discusses tcMenu issues. But most likely either the code is blocking, or crashing within that call. Have you reused or misconfigured a pin?

I think you're better off trying this code in a simple test harness first and getting the code working in such a way that you could use it in the dashboard. IE the sensor reading code and not blocking.

**Reply by davetcc** [reply]
Also as per docs don’t call delay in any run loops.

**Reply by adhossain** [reply]
ok.

**Reply by adhossain** [reply]
yes. the code works fine in simple form. if put in loop.

**Reply by davetcc** [reply]
You're going to need to debug it to see where it is failing, to see if there is some interaction or similar. I normally use a hardware debugger for that, but if you don't have one then put print statements all around that code and see where the trail ends. I suspect either a pin is being reassigned somewhere or something is blocking. I would compare your menu project carefully with the test harness.

I mean just looking at the first block `} while (Ultrasonic_Sensor.read() == 0xff);` you can see it would block until there were data. Given this is ESP32, you could either do that on another thread if you're used to multithreaded programming and understand the memory implications, or recode it to be asynchronous and not block.
